src/common/options.cc

```diff
@@ -1632,6 +1632,12 @@ std::vector<Option> get_global_options() {
     .set_default(0)
     .set_description(""),
 
+    Option("mon_warn_pg_scrubs_overdue_ratio", Option::TYPE_FLOAT, Option::LEVEL_ADVANCED)
+    .set_default(.25)
+    .set_description("Warn when any osd has this percentage of pgs or this percentage of all pgs with scrubs overdue. "
+                     "The value of osd_max_scrubs is also accounted for.")
```

Comment:
Warn when this fraction of PGs have overdue scrubs.

Reply:
An alternative way to approach this would be to specify a ratio of the scrub interval, and warn when any PGs are haven't scrubbed in this long.  So, for example, if you scrub every 2 weeks, and changed it to scrub every week, and your ratio was .5, then you'd have  warning that all PGs are more than 50% past due.  That number would eventually drain down to 0 and the warning would go away.

One concern with the way this is now is that if you have one pool that doesn't scrub for some reason, you could have most PGs getting scrubbed fine but a minority (say, 5%) that is perpentually behind but not see a health warning about it...

Reply:
Also, since the available information here is all in PGMap on the mgr (we just need to look at last_scrub and last_deep_scrub), I'm not sure we should need any changes to osd_stat_t at all?

Reply:
@liewegas I'm not sure how a PG would not get scrubbed except due to a long/stuck recovery.  Normally, the first PG to scrub is the one most overdue.  My new code could violate this if you keep requesting scrubs to always starve the overdue scheduled ones.  However, this change fixes a problem with scheduled scrubs making requested scrubs wait a long time.

In the current implementation a deadline is stored in the ``sched_scrub_pg`` queue (actually a set<>) when a pg registers its next scrub.  Changing the ``osd_scrub_max_interval`` config or the pool's ``scrub_max_interval`` doesn't affect a pg until it next scrubs and registers again.  The reason I don't use the ``last_scrub`` is that the ``deadline`` is used to determine if a pg is overdue and waiting its turn to run.  Also, the most recent change I made requires knowing if a pg is a requested scrub (must scrub).  I don't think that ``scrubber.must_scrub`` are available in MGR to adjust for.  Deep scrub is analogous.

Reply:
The last_[deep_]scrub fields are completely independent of the admin/forced scrubs.. which I think is a good thing?

The warning could happen up if no[deep-]scrub is set for a long time, or if scrub is just too slow to keep up with the scrub interval.  So if the interval is 1 week, plus a rnadomization factor of +/- 25%, then we could default the warning threshold to 1.5 or 2x the interval so it would trigger if the pg was due to scrub for a significant portion of the next interval.

I think that could be done without touching the OSD code at all, and just be implemented in the PGMap health method?

Reply:
I don't think that is a good thing.  I like being able to notice that requested scrubs are backing things up themselves or in combination with the overdue scheduled scrubs.  The ``deadline`` is modified when you request a scrub which isn't reflected in the ``last_[deep_]scrub`` fields at all.   To maintain what I want will require the OSD code changes.

The ``osd_[deep_]scrub_interval_randomize_ratio`` is more subtle.  It involves the ``osd_scrub_min_interval`` NOT the ``deadline`` which is set to ``now + scrub_max_interval``.  So the randomization won't affect the current code unless min/max are too close together or randomization ratio is very large.  The defaults are min 1 day, random 50% so up to an extra 1/2 day and max is 1 week.

If the cluster is brought up and some initial pools created, all PGs have about the same ``deadline``, under load all week we will pass it on all PGs at exactly the 1 week mark.  If not under load every day each PG will scrub and set the deadline 1 week out.   I think that under normal operations, scrubs will tend to spread themselves out and even under load PGs will become overdue over a period of time.